torch/_numpy/random.py

```diff
@@ -98,15 +116,33 @@ def normal(loc=0.0, scale=1.0, size=None):
         size = ()
     dtype = _dtypes_impl.default_dtypes().float_dtype
     values = torch.empty(size, dtype=dtype).normal_(loc, scale)
-    return array_or_scalar(values, return_scalar=size is None)
+    return array_or_scalar(values, return_scalar=size == ())
 
 
 @deco_stream
-@normalizer
 def shuffle(x: ArrayLike):
-    perm = torch.randperm(x.shape[0])
-    xp = x[perm]
-    x.copy_(xp)
+    # no @normalizer because we need to shuffle e.g. lists in-place
+    from ._ndarray import ndarray
+
+    if isinstance(x, torch.Tensor):
+        _shuffle_tensor_inplace(x)
+    elif isinstance(x, ndarray):
+        _shuffle_tensor_inplace(x.tensor)
+    else:
+        # untyped code path. Apply the permutation in-place
+        # cf https://stackoverflow.com/questions/16501424/algorithm-to-apply-permutation-in-constant-memory-space/41472796#41472796
+        perm = torch.randperm(len(x)).tolist()
+        for i in range(len(x)):
+            idx = perm[i]
+            while idx < i:
+                idx = perm[idx]
+            x[i], x[idx] = x[idx], x[i]
```

Comment:
If the input is not a tensor let's just raise a NotImplementedError.

Reply:
After doing so, we can pretty much leave the implementation as it was modulo an early if

Reply:
That would make our lives easier indeed, but it's kinda hard to explain. Previously, the random stream is either that of numpy or py torch, depending on the global flag. 

With a graph break, the stream depends on the argument type.

Reply:
If we graphbreak the stream is always going to be that of NumPy, so that's consistent all across

Reply:
Under dynamo, yes. Do not care about eager at all?

Reply:
Not to this extent. Any place were we break via a `NotImplementedError` is going to differ between eager and compiled, but that's alright.